### ISSUE TITLE
Errors when using translate directive with dynamic value and messageformat pluralization

### ISSUE THREAD

**Issue by tqmukas** [angular-translate/angular-translate#789]
Please see the plunker example:
http://embed.plnkr.co/RhuPW6cqpfsSuBS1R45a/preview

As far as I understand it's because ```angular-translate.js``` tries to submit ```translate-values``` while those are not ready. ```messageformat.js ``` library receives empty values and throws an error.

**Comment by morgoth** [comment]
The same issue here after upgrading `angular-translate` to 2.5.2 (on 2.4.2 it works fine)

**Comment by einyalynn** [comment]
don't know if this is helpful for anyone in the short term, but if you start forcing interpolation combined with using the translate-value-* directive you can get around this.

Here's an updated plunker of the original bug:
http://embed.plnkr.co/Qg3OuIlcGZ6G4Kij6tz6/preview

**Comment by knalli** [comment]
@gontard What would be the bad practice to avoid that? We cannot avoid this call with an empty object, because this is a valid process in angular-translate (empty interpolation params). 

An actual translation is required.

**Comment by DieterAtWork** [comment]
@einyalynn , this doesn't work here on the latest release. When using translate-value-*, it isn't translated at all.

**Comment by phw** [comment]
I can confirm this issue. It happens to me when I upgrade to angular-translate 2.5.2, version 2.4.2 works fine. Both times using messageformat 0.1.8

**Comment by RBLU** [comment]
also seeing this issue when updating to 2.5.2 (with angular 1.3.10). reverted to angular-translate 2.4.2 - there we don't experience this issue.

**Comment by knalli** [comment]
This issue is still open because.. well, there is still no solution.

> In my case stringcontains the message format pattern but the interpolateParams are just {} so MessageFormat fails to evaluated the access to the attribute.

So.. should we try/catch it only and wait for a change of the interpolate params? Because without a bad duck check (is there a curly brace in the string), we actually do not know whether the translation is fine for MF or not.

**Comment by lboynton** [comment]
Not sure if this is relevant, but for me `interpolate()` is being called twice. The first time `interpolateParams` is an empty object, the second time `interpolateParams` is correct; it contains the object passed into `translate-values=""`. Should the first call not be happening?

**Comment by tspaeth** [comment]
@lboynton On which version are you? Latest 2.6.x?

**Comment by lboynton** [comment]
@tspaeth 2.5.2, but I also get the same problem with 2.6.1. 2.4.2 does not throw this error, like other people have mentioned. Using angular 1.3.14 and messageformat 0.1.8.

**Comment by tspaeth** [comment]
@lboynton could you check your setup against the just created PR locally? 
Not sure if it will fix it, but in that case, the value-attributes are initially present on the first $translate call already... just a try.

@knalli I'm not pretty sure if it has any side effects for us. Could you have a cross-check please?

**Comment by lboynton** [comment]
@tspaeth I did test out your changes and they didn't seem to fix the problem for me. I intended to do some more testing but haven't got round to that yet.

**Comment by tspaeth** [comment]
@lboynton Today I've commented on it again (see #948 comments) to provide a fix if you're using translate-values. Did you already test it included?

**Comment by ombr** [comment]
Is there any news on this issue ?

**Comment by knalli** [comment]
@ombr : Current `canary` contains 2533f2d -- it would be helpful if you can confirm that one would fix this for you.

**Comment by knalli** [comment]
@ombr , @lboynton can you confirm the bugfix?

**Comment by andreialecu** [comment]
I just tried the latest canary but couldn't get it to work at all. The main .js was broken for my app. I got an error saying `Unknown provider `$translateSanitizationProvider`` which I was able to resolve by reordering the .js files in Gruntfile.js, and putting `src/service/sanitization.js` right after `src/translate.js` then rebuilding.

Anyway, after getting canary working, I verified that this bug still isn't fixed and I'm getting the same errors in the console: `Error: MessageFormat: 'count' isnt a number.`.

In my project, the `count` value is populated dynamically from a REST service, so it isn't available right away.

I have edited the original plunker here: 
http://plnkr.co/edit/UZlzqvmx9JSKQyn3o6Bo?p=preview

The error still occurs in the part I highlighted, but the others seem fine. Changing ng-show to ng-if seems to fix it, but I'm not sure I can do that in my code without breaking other things. Any ideas?

**Comment by andreialecu** [comment]
A little bit offtopic, regarding the $translateSanitizationProvider issue: here's the broken/original canary version:
http://plnkr.co/edit/hiyJbo49bkLpR4pTrMgt?p=preview

This is the broken .js as generated by `grunt build`:
https://gist.githubusercontent.com/andreialecu/70ee6839d72221993d3e/raw/95dcc0bfba72602d94188df6c0eca1be904ed5d7/angular-translate-broken.js

Paging @marklagendijk and @knalli regarding https://github.com/angular-translate/angular-translate/commit/12dbc5754c8315d41144794918f1ec8ca34c601e

**Comment by knalli** [comment]
@andreialecu  Gotcha. Thank you. Fixed in d298eb2, working example http://plnkr.co/edit/pcgamKitVbpyfyzfz88e?p=preview

@andreialecu are you able to make a working demo of your setup based/forked on this version?

**Comment by andreialecu** [comment]
@knalli I have updated the plunker originally posted at the top of this thread to the latest canary.

http://plnkr.co/edit/UZlzqvmx9JSKQyn3o6Bo?p=preview

There's a certain use case where the exception is still logged, even though everything works.

Is there any way to get rid of it? The only way I can see that works is to add `ng-if` to the element so it isn't created if that value doesn't exist, but it is a bit cumbersome to add it to all my elements. I think the exception should simply just be ignored, as with vanilla (non MF) angular-translate.

Also, `translate-value-*` doesn't seem to work at all with MF, is this expected?

**Comment by knalli** [comment]
@andreialecu Thank you.

I've forked and updated your latest plnkr with a patch: http://plnkr.co/edit/U3Uw8M61MzgOkEpYc2R6?p=preview

Following notes:

1. MessageFormat looks like the actual type of the variable. Due technical limitations, we (based on the data given by DOM/AngularJS) cannot say if a value is a string or number. I've added a small duck type check provided here: bbc1cbe
2. The expressions use uppercase value keys. The issue with this is AngularJS does not provide this case sensitive information (i.e. a `translate-value-AMOUNT` will be internally treated as `translateValueAmount` where we skip the first 2 words and de-capitalize the actual value key to `amount`). If you use lowercase expression variables, you will getting no more warnings.
3. There was an issue with `translate-value-*` was not being available the first time when, but on the next update. Fixed in 98e8279
